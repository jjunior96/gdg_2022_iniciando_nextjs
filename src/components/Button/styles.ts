import { darken, lighten } from 'polished';
import styled, { css, DefaultTheme } from 'styled-components';

import { ButtonProps } from '.';

const containerModifiers = {
  disabled: (theme: DefaultTheme) => css`
    cursor: not-allowed;
    background-color: ${theme.colors.gray_300};

    &:hover {
      background-color: ${theme.colors.gray_300};

      box-shadow: none;
    }
  `,

  small: (theme: DefaultTheme) => css`
    height: 2.4rem;

    font-size: ${theme.font.sizes.xxsmall};
  `,

  medium: (theme: DefaultTheme) => css`
    height: 3.2rem;

    font-size: ${theme.font.sizes.xsmall};
  `,

  large: (theme: DefaultTheme) => css`
    height: 5.6rem;

    font-size: ${theme.font.sizes.xsmall};
  `,

  minimal: (theme: DefaultTheme) => css`
    background: none;

    color: ${theme.colors.green};

    font-weight: ${theme.font.normal};

    transition: color 0.2s;

    &:hover {
      color: ${darken(0.1, theme.colors.green)};
    }
  `
};

const variants = {
  outline: (theme: DefaultTheme) => css`
    background-color: transparent;
    color: ${theme.colors.primary};
    border: 2px solid ${theme.colors.primary};

    &:hover {
      background-color: transparent;
      color: ${lighten(0.1, theme.colors.primary)};

      border-color: ${lighten(0.1, theme.colors.primary)};
    }
  `
};

export const Container = styled.button<ButtonProps>`
  ${({ theme, color, size, minimal, disabled, variant }) => css`
    width: 100%;

    color: ${theme.colors.white};
    background-color: ${theme.colors[color!]};

    display: flex;
    justify-content: center;
    align-items: center;

    border-radius: ${theme.border.radius};
    border: 0;

    padding: 0 ${theme.spacings.xsmall};

    font-weight: ${theme.font.bold};
    font-size: ${theme.font.sizes.xsmall};

    transition: background-color 0.3s, box-shadow 0.2s;

    outline: none;

    &:focus {
      outline: 3px solid ${lighten(0.4, theme.colors.green)};
    }

    ${!minimal &&
    css`
      &:hover {
        background-color: ${lighten(0.05, theme.colors[color!])};

        box-shadow: ${theme.box.shadow};
      }
    `}

    > svg {
      margin-right: ${theme.spacings.xxsmall};
    }

    ${!!size && containerModifiers[size](theme)};

    ${!!minimal && containerModifiers.minimal(theme)};
    ${!!variant && variants[variant!](theme)};
    ${disabled && containerModifiers.disabled(theme)};

    ${color === 'red' &&
    css`
      color: ${theme.colors.white};
      border-color: ${theme.colors.red};

      &:hover {
        border-color: ${theme.colors.red};

        box-shadow: ${theme.box.shadow};
      }
    `}
  `}
`;
