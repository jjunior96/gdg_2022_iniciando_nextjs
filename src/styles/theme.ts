export default {
  grid: {
    container: '96rem',
    gutter: '3.2rem'
  },
  components: {
    sidebar: '20rem',
    header: '8rem'
  },
  border: {
    radius: '0.4rem'
  },
  box: {
    shadow: '0 8px 8px rgba(0, 0, 0, 0.2)'
  },
  font: {
    family:
      "Montserrat, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif",
    light: 300,
    normal: 400,
    bold: 600,
    sizes: {
      xxxxsmall: '0.8rem',
      xxxsmall: '1.2rem',
      xxsmall: '1.4rem',
      xsmall: '1.6rem',
      small: '2rem',
      medium: '2.4rem',
      large: '3.2rem',
      xlarge: '4rem',
      xxlarge: '4.8rem',
      xxxlarge: '5.6rem',
      huge: '6.4rem'
    }
  },
  colors: {
    primary: '#24A66B',
    secundary: '#5855E9',
    white: '#FAFAFA',
    gray_50: '#f0efeb',
    gray_100: '#D9D9D9',
    gray_200: '#91AAB4',
    gray_300: '#3E606F',
    gray_450: '#193441',
    gray_500: '#343a40',
    yellow_300: '#F1C40F',
    purple_100: '#7F57DF',
    purple_300: '#5855E9',
    green: '#24A66B',
    red: '#E63737',
    success: '#24A66B',
    error: '#E63737',
    info: '#189CEF',
    black: '#0D0D0D'
  },
  spacings: {
    xxxsmall: '0.8rem',
    xxsmall: '1rem',
    xsmall: '1.6rem',
    small: '2.4rem',
    medium: '3.2rem',
    large: '4.0rem',
    xlarge: '4.8rem',
    xxlarge: '5.6rem'
  },
  layers: {
    base: 10,
    menu: 20,
    overlay: 30,
    modal: 40,
    alwaysOnTop: 50
  },
  modal: {
    sizes: {
      small: '40rem',
      medium: '48rem',
      large: '54rem'
    }
  },
  transition: {
    default: '0.3s ease-in-out',
    fast: '0.1s ease-in-out'
  }
} as const;
