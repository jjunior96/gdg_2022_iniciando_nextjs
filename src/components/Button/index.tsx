import React, { AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react';

import Spinner from 'components/Spinner';

import * as S from './styles';

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>;

export type ButtonVariantsProps = 'outline';

type Props = {
  size?: 'small' | 'medium' | 'large';
  color?: 'green' | 'red';
  children: React.ReactNode;
  minimal?: boolean;
  isLoading?: boolean;
  disabled?: boolean;
  icon?: React.ReactNode;
  as?: React.ElementType;
  variant?: ButtonVariantsProps;
};

export type ButtonProps = ButtonTypes & Props;

const Button = ({
  children,
  size = 'large',
  color = 'green',
  minimal = false,
  isLoading = false,
  disabled = false,
  variant,
  icon,
  ...rest
}: ButtonProps) => {
  return (
    <S.Container
      size={size}
      color={color}
      minimal={minimal}
      variant={variant}
      disabled={disabled || isLoading}
      {...rest}
    >
      {!!icon && icon}
      {isLoading ? <Spinner size="large" /> : children}
    </S.Container>
  );
};

export default Button;
