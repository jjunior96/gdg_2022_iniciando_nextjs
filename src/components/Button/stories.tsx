import { AiOutlineHeart as HeartIcon } from 'react-icons/ai';
import { FiLogIn as LoginIcon } from 'react-icons/fi';

import { Story, Meta } from '@storybook/react/types-6-0';

import Button, { ButtonProps } from '.';

export default {
  title: 'Button',
  component: Button,
  decorators: [
    (Story) => (
      <div
        style={{
          width: '40rem',
          margin: '0 auto',
          marginTop: '1rem',
          display: 'flex',
          justifyContent: 'center',
          alignItems: 'center'
        }}
      >
        <Story />
      </div>
    )
  ],
  parameters: {
    backgrounds: {
      default: 'white'
    }
  },
  argTypes: {
    size: {
      control: {
        type: 'select',
        options: ['small', 'medium', 'large']
      }
    },
    as: {
      type: ''
    },
    href: {
      type: ''
    },
    icon: {
      type: ''
    }
  }
} as Meta;

export const Default: Story<ButtonProps> = (args) => <Button {...args} />;

Default.args = {
  children: 'Enter'
};

export const withIcon: Story<ButtonProps> = (args) => <Button {...args} />;

withIcon.args = {
  children: 'Enter',
  icon: <LoginIcon />
};

export const withLoading: Story<ButtonProps> = (args) => <Button {...args} />;

withLoading.args = {
  isLoading: true,
  children: 'Enter'
};

export const outlined: Story<ButtonProps> = (args) => <Button {...args} />;

outlined.args = {
  children: 'Enter',
  variant: 'outline'
};

export const minimal: Story<ButtonProps> = (args) => <Button {...args} />;

minimal.args = {
  children: 'Enter',
  minimal: true,
  icon: <HeartIcon />
};

export const asLink: Story<ButtonProps> = (args) => <Button {...args} />;

asLink.args = {
  children: 'Enter',
  as: 'a',
  href: '/link'
};
