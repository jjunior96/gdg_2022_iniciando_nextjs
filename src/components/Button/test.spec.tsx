import { AiOutlineHeart as HeartIcon } from 'react-icons/ai';

import { screen } from '@testing-library/react';

import { renderWithTheme } from 'utils/tests/__config__/helpers';

import theme from 'styles/theme';

import Button from '.';

describe('<Button />', () => {
  it('should render a button default', () => {
    const { container } = renderWithTheme(<Button>Enter</Button>);

    expect(screen.getByText(/Enter/i)).toHaveStyle({
      height: '5.6rem',
      'background-color': theme.colors.green,
      color: theme.colors.white
    });

    expect(container).toMatchSnapshot();
  });

  it('should render a button with size small', () => {
    renderWithTheme(<Button size="small">Enter</Button>);

    expect(screen.getByText(/Enter/i)).toHaveStyle({
      height: '2.4rem'
    });
  });

  it('should render a button with size medium', () => {
    renderWithTheme(<Button size="medium">Enter</Button>);

    expect(screen.getByText(/Enter/i)).toHaveStyle({
      height: '3.2rem'
    });
  });

  it('should render a button with size large', () => {
    renderWithTheme(<Button size="large">Enter</Button>);

    expect(screen.getByText(/Enter/i)).toHaveStyle({
      height: '5.6rem'
    });
  });

  it('should render a button with color green', () => {
    renderWithTheme(<Button color="green">Enter</Button>);

    expect(screen.getByText(/Enter/i)).toHaveStyle({
      'background-color': theme.colors.green
    });
  });

  it('should render a button with color red', () => {
    renderWithTheme(<Button color="red">Enter</Button>);

    expect(screen.getByText(/Enter/i)).toHaveStyle({
      'background-color': theme.colors.red
    });
  });

  it('should render a minimal version', () => {
    renderWithTheme(<Button minimal>Enter</Button>);

    expect(screen.getByRole('button', { name: /Enter/i })).toHaveStyle({
      background: 'none',
      color: theme.colors.green
    });
  });

  it('should render a button as a link (<a/>)', () => {
    renderWithTheme(
      <Button color="red" as="a" href="/">
        Enter
      </Button>
    );

    expect(screen.getByRole('link', { name: /Enter/i })).toHaveAttribute(
      'href',
      '/'
    );
  });

  it('should render a spinner loading when isLoading passed', () => {
    renderWithTheme(<Button isLoading>Enter</Button>);

    expect(screen.getByRole('button')).toHaveStyle({
      background: theme.colors.gray_300,
      cursor: 'not-allowed'
    });

    // Verifica se o Spinner estÃ¡ aparecendo
    expect(screen.getByTestId('spinner')).toBeInTheDocument();
  });

  it('should render without icon when not passed', () => {
    renderWithTheme(<Button>Enter</Button>);

    const iconComponent = screen.queryByTestId(/empty-icon/);

    expect(iconComponent).not.toBeInTheDocument();
  });

  it('should render with icon when passed', () => {
    renderWithTheme(
      <Button icon={<HeartIcon data-testid="button-icon" />}>Enter</Button>
    );

    const iconComponent = screen.queryByTestId(/button-icon/i);

    expect(iconComponent).toBeInTheDocument();
  });

  it('should render with outline variant when passed', () => {
    renderWithTheme(<Button variant="outline">Enter</Button>);

    const buttonComponent = screen.getByText(/Enter/i);

    expect(buttonComponent).toHaveStyle({
      'background-color': 'transparent',
      color: theme.colors.primary,
      border: `2px solid ${theme.colors.primary}`
    });
  });
});
