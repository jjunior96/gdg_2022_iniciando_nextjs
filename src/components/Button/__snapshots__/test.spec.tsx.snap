// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render a button default 1`] = `
.c0 {
  width: 100%;
  color: #FAFAFA;
  background-color: #24A66B;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0.4rem;
  border: 0;
  padding: 0 1.6rem;
  font-weight: 600;
  font-size: 1.6rem;
  -webkit-transition: background-color 0.3s,box-shadow 0.2s;
  transition: background-color 0.3s,box-shadow 0.2s;
  outline: none;
  height: 5.6rem;
  font-size: 1.6rem;
}

.c0:focus {
  outline: 3px solid #aaecce;
}

.c0:hover {
  background-color: #29bb79;
  box-shadow: 0 8px 8px rgba(0,0,0,0.2);
}

.c0 > svg {
  margin-right: 1rem;
}

<div>
  <button
    class="c0"
    color="green"
  >
    Enter
  </button>
</div>
`;
